### ISSUE TITLE
FCF formal coerced to borrowed within body of coforall

### ISSUE THREAD

**Issue by dlongnecke-cray** [chapel-lang/chapel#16095]
```chapel
iter p1(worker: func(int, void)) {
  coforall loc in Locales do on loc {
    yield p2(worker);
  }
}

iter p2(worker: func(int, void)) { worker(0); yield 0; }

proc doWork(x: int) {
  writeln(x);
}

proc test() {
  var tmp = p1(doWork);
  writeln(tmp);
}
test();
```

```
test2.chpl:1: In iterator 'p1':
test2.chpl:3: error: unresolved call 'p2(borrowed chpl__fcf_type_int64_t_void)'
test2.chpl:7: note: this candidate did not match: p2(worker: func(int, void))
test2.chpl:3: note: because call actual argument #1 with type borrowed chpl__fcf_type_int64_t_void
test2.chpl:7: note: is passed to formal 'worker: shared chpl__fcf_type_int64_t_void'
```

It seems that first class function formals are being incorrectly coerced to borrowed when copied into `coforall` loop bodies for one reason or another.

**Comment by bradcray** [comment]
FCFs are implemented as `owned` classes under the hood, so I believe what you're seeing here is consistent with what happens more generally when `owned` classes enter a parallel region: a shadow variable that borrows from them is taken.  Meanwhile, I think the formal type `func()` is probably hard-coded to `owned` causing the mismatch.  A fix _might_ be to treat formal `func` types as being generic w.r.t. the ownership type of the FCF, but this also raises questions for me about whether it's OK to borrow a FCF without consequence and potentially have multiple things call the same object simultaneously (which relates a bit to whether FCF's capture variables or not, issue #11506.

It would probably be worth adding this to #8351 FCF epic as well...

**Comment by dlongnecke-cray** [comment]
Are you sure FCFs are `owned` classes? I could have sworn I modified them to be `shared` in order to fix the FCF memory leak (this would also be consistent with the error message in the OP that mentions the formal as `shared`).

**Comment by bradcray** [comment]
Ah, no... not at all...  I was misremembering how we fixed that leak (and forgot that you had implemented it, and obviously didn't stare at the error message very closely).

That said, I think the rest of the diagnosis stands, as `shared` classes currently also turn into `borrowed` when they cross a parallel section:

```
class C {
}

var myC = new shared C();

coforall i in 1..10 {
  writeln(myC.type:string);
}
```

**Comment by dlongnecke-cray** [comment]
Yeah that diagnosis seems reasonable. I just am not entirely sure if the behavior is intentional or not. I can understand why the coercion to borrowed might be the case for owned, but AFAIK shared can be copied freely so it seems strange not to copy it into the task (as I would hope we also do for unmanaged).

I guess one workaround would be to detect a FCF type and then construct a shared from the borrow? We probably want to have formal FCFs remain shared incase one is assigned to e.g. a structure field.

**Comment by bradcray** [comment]
I agree that that's a fair question ("Should `shared` classes be `borrowed` when they cross task boundaries or remain `shared`?") and that if the answer was "remain `shared`" then that would be a simple fix for this issue.  I'd fork that question off into a topic of its own (I don't know whether it was intentional or discussed or not either).

**Comment by dlongnecke-cray** [comment]
Thanks, will do!
